### ISSUE TITLE
CrashLoopBackoff on quickstart

### ISSUE THREAD

**Issue by instantlinux** [CrunchyData/postgres-operator#1901]
**Describe the bug**
Quickstart fails in default configuration, or with any other storage setting that I attempt, with this error:
```
time="2020-09-21T05:07:32Z" level=error msg="Error in pgoconfig: check pgo.yaml: PrimaryStorage setting is invalid: \"dynamic\" is not defined" func="internal/apiserver.Initialize()" file="internal/apiserver/root.go:134" version=4.4.1
```
**To Reproduce**
1. `kubectl create namespace pgo`
2. `kubectl apply -f https://raw.githubusercontent.com/CrunchyData/postgres-operator/v4.4.1/installers/kubectl/postgres-operator.yml`

Literally, the two-step quickstart.

**Expected behavior**
A working environment.

**Please tell us about your environment:**

* Operating System: Ubuntu 18.04
* Where is this running ( Local, Cloud Provider): Local
* Storage being used (NFS, Hostpath, Gluster, etc): hostpath
* Container Image Tag: 4.4.1
* PostgreSQL Version: default
* Platform (Docker, Kubernetes, OpenShift): kubernetes
* Platform Version: 1.15.3

**Additional context**
I'm a newbie so this may be a repeat request--but it shouldn't take me two hours of struggling to conclude that, basically, the quickstart isn't quick and requires a bug report.

Logs:
```
$ kubectl logs  -n pgo --context=sudo -f postgres-operator-66cc8bd589-9nbdf apiserver
time="2020-09-21T05:20:24Z" level=info msg="debug flag set to false"
time="2020-09-21T05:20:29Z" level=info msg="postgres-operator apiserver starts" func="main.main()" file="apiserver.go:109" version=4.4.1
time="2020-09-21T05:20:29Z" level=info msg="Pgo Namespace is [pgo]" func="internal/apiserver.Initialize()" file="internal/apiserver/root.go:107" version=4.4.1
time="2020-09-21T05:20:29Z" level=info msg="InstallationName is [devtest]" func="internal/apiserver.Initialize()" file="internal/apiserver/root.go:114" version=4.4.1
time="2020-09-21T05:20:29Z" level=info msg="apiserver starts" func="internal/apiserver.Initialize()" file="internal/apiserver/root.go:126" version=4.4.1
time="2020-09-21T05:20:29Z" level=info msg="loading PermMap with 60 Permissions\n" func="internal/apiserver.InitializePerms()" file="internal/apiserver/perms.go:189" version=4.4.1
time="2020-09-21T05:20:29Z" level=info msg="Config: pgo-config ConfigMap found, using config files from the configmap" func="internal/config.getRootPath()" file="internal/config/pgoconfig.go:753" version=4.4.1
time="2020-09-21T05:20:29Z" level=error msg="Error in pgoconfig: check pgo.yaml: PrimaryStorage setting is invalid: \"dynamic\" is not defined" func="internal/config.(*PgoConfig).GetConfig()" file="internal/config/pgoconfig.go:553" version=4.4.1
time="2020-09-21T05:20:29Z" level=error msg="Error in pgoconfig: check pgo.yaml: PrimaryStorage setting is invalid: \"dynamic\" is not defined" func="internal/apiserver.Initialize()" file="internal/apiserver/root.go:134" version=4.4.1
time="2020-09-21T05:20:29Z" level=error msg="error in Pgo configuration" func="internal/apiserver.Initialize()" file="internal/apiserver/root.go:135" version=4.4.1
```

**Comment by instantlinux** [comment]
Log from the ansible pgo-deploy:
```
(python_env) richb@mckinley:~/docker/k8s$ kubectl logs  -n pgo --context=sudo -f pgo-deploy-xz5mq 

PLAY [Deploy Crunchy PostgreSQL Operator] **************************************

TASK [Gathering Facts] *********************************************************
ok: [localhost]

TASK [pgo-preflight : include_tasks] *******************************************
included: /ansible/roles/pgo-preflight/tasks/vars.yml for localhost

TASK [pgo-preflight : Check for "/conf/values.yaml"] ***************************
ok: [localhost]

TASK [pgo-preflight : fail] ****************************************************
skipping: [localhost]

TASK [pgo-preflight : include_vars] ********************************************
ok: [localhost]

TASK [pgo-preflight : fail] ****************************************************
skipping: [localhost]

TASK [pgo-preflight : Check Operating System] **********************************
changed: [localhost]

TASK [pgo-preflight : assert] **************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "All assertions passed"
}

TASK [pgo-preflight : assert] **************************************************
skipping: [localhost]

TASK [pgo-preflight : assert] **************************************************
skipping: [localhost]

TASK [pgo-preflight : assert] **************************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "All assertions passed"
}

TASK [pgo-preflight : include_tasks] *******************************************
skipping: [localhost]

TASK [pgo-preflight : include_tasks] *******************************************
included: /ansible/roles/pgo-preflight/tasks/check_kubernetes.yml for localhost

TASK [pgo-preflight : Check if the kubectl command is installed] ***************
changed: [localhost]

TASK [pgo-preflight : Ensure kubectl is installed] *****************************
ok: [localhost] => {
    "changed": false, 
    "msg": "All assertions passed"
}

TASK [pgo-preflight : include_tasks] *******************************************
included: /ansible/roles/pgo-preflight/tasks/check_vars.yml for localhost

TASK [pgo-preflight : Check if mandatory variables are defined] ****************
skipping: [localhost] => (item=pgo_operator_namespace) 
skipping: [localhost] => (item=pgo_installation_name) 
skipping: [localhost] => (item=pgo_admin_username) 
skipping: [localhost] => (item=pgo_admin_password) 
skipping: [localhost] => (item=pgo_admin_role_name) 
skipping: [localhost] => (item=pgo_admin_perms) 
skipping: [localhost] => (item=ccp_image_prefix) 
skipping: [localhost] => (item=ccp_image_tag) 
skipping: [localhost] => (item=pgo_image_prefix) 
skipping: [localhost] => (item=pgo_image_tag) 
skipping: [localhost] => (item=disable_auto_failover) 
skipping: [localhost] => (item=badger) 
skipping: [localhost] => (item=metrics) 
skipping: [localhost] => (item=archive_mode) 
skipping: [localhost] => (item=archive_timeout) 
skipping: [localhost] => (item=db_password_length) 
skipping: [localhost] => (item=create_rbac) 
skipping: [localhost] => (item=db_port) 
skipping: [localhost] => (item=db_replicas) 
skipping: [localhost] => (item=db_user) 
skipping: [localhost] => (item=backrest_storage) 
skipping: [localhost] => (item=backup_storage) 
skipping: [localhost] => (item=primary_storage) 
skipping: [localhost] => (item=replica_storage) 
skipping: [localhost] => (item=pgo_client_version) 
skipping: [localhost] => (item=pgbadgerport) 
skipping: [localhost] => (item=exporterport) 
skipping: [localhost] => (item=scheduler_timeout) 
skipping: [localhost] => (item=namespace_mode) 
skipping: [localhost] => (item=reconcile_rbac) 

TASK [pgo-operator : Set output directory fact] ********************************
ok: [localhost]

TASK [pgo-operator : Ensure output directory exists] ***************************
changed: [localhost]

TASK [pgo-operator : include_tasks] ********************************************
skipping: [localhost] => (item=openshift_auth.yml) 
skipping: [localhost] => (item=openshift.yml) 

TASK [pgo-operator : include_tasks] ********************************************
included: /ansible/roles/pgo-operator/tasks/kubernetes_auth.yml for localhost
included: /ansible/roles/pgo-operator/tasks/kubernetes.yml for localhost

TASK [pgo-operator : Set the Kubernetes Context] *******************************
skipping: [localhost]

TASK [pgo-operator : Get Namespace Details] ************************************
changed: [localhost]

TASK [pgo-operator : Create PGO Namespace] *************************************
skipping: [localhost]

TASK [pgo-operator : include_tasks] ********************************************
included: /ansible/roles/pgo-operator/tasks/certs.yml for localhost

TASK [pgo-operator : Ensure directory exists for local self-signed TLS certs.] ***
ok: [localhost]

TASK [pgo-operator : Generate RSA Key] *****************************************
changed: [localhost]

TASK [pgo-operator : Generate CSR] *********************************************
changed: [localhost]

TASK [pgo-operator : Generate Self-signed Certificate] *************************
changed: [localhost]

TASK [pgo-operator : Ensure /tmp/.pgo/pgo Directory Exists] ********************
ok: [localhost]

TASK [pgo-operator : Copy certificates to /tmp/.pgo/pgo] ***********************
changed: [localhost]

TASK [pgo-operator : Copy keys to /tmp/.pgo/pgo] *******************************
changed: [localhost]

TASK [pgo-operator : Use kubectl or oc] ****************************************
ok: [localhost]

TASK [pgo-operator : include_tasks] ********************************************
included: /ansible/roles/pgo-operator/tasks/namespace.yml for localhost

TASK [pgo-operator : Namespace List] *******************************************
ok: [localhost]

TASK [pgo-operator : Create Watched Namespaces] ********************************
skipping: [localhost] => (item=pgo) 

TASK [pgo-operator : Create Watched Namespaces (Reconcile RBAC)] ***************
skipping: [localhost] => (item=pgo) 

TASK [pgo-operator : Label Watched Namespaces (Reconcile RBAC)] ****************
skipping: [localhost] => (item=pgo) 

TASK [pgo-operator : Cleanup Local Namespace Target RBAC] **********************
skipping: [localhost] => (item=pgo) 

TASK [pgo-operator : Create Local Namespace Target RBAC] ***********************
skipping: [localhost] => (item=pgo) 

TASK [pgo-operator : Template Local Namespace RBAC] ****************************
skipping: [localhost]

TASK [pgo-operator : Cleanup Local Namespace Reconcile RBAC] *******************
skipping: [localhost] => (item=pgo) 

TASK [pgo-operator : Create Local Namespace Reconcile RBAC] ********************
skipping: [localhost] => (item=pgo) 

TASK [pgo-operator : include_tasks] ********************************************
included: /ansible/roles/pgo-operator/tasks/crds.yml for localhost

TASK [pgo-operator : Check if PGCluster CRD Is Installed] **********************
changed: [localhost]

TASK [pgo-operator : Create PGClusters CRD] ************************************
skipping: [localhost]

TASK [pgo-operator : Check if PGPolicies CRD Is Installed] *********************
changed: [localhost]

TASK [pgo-operator : Create PGPolicies CRD] ************************************
skipping: [localhost]

TASK [pgo-operator : Check if PGReplicas CRD Is Installed] *********************
changed: [localhost]

TASK [pgo-operator : Create PGReplicas CRD] ************************************
skipping: [localhost]

TASK [pgo-operator : Check if PGTasks CRD Is Installed] ************************
changed: [localhost]

TASK [pgo-operator : Create PGTasks CRD] ***************************************
skipping: [localhost]

TASK [pgo-operator : Template PGO Admin Credentials] ***************************
changed: [localhost]

TASK [pgo-operator : Check PGO Admin Credentials] ******************************
changed: [localhost]

TASK [pgo-operator : Create PGO Admin Credentials] *****************************
skipping: [localhost]

TASK [pgo-operator : Template PGO Admin Role & Permissions] ********************
changed: [localhost]

TASK [pgo-operator : Check PGO Admin Role & Permissions] ***********************
changed: [localhost]

TASK [pgo-operator : Create PGO Admin Role & Permissions] **********************
skipping: [localhost]

TASK [pgo-operator : Template PGO Service Account] *****************************
changed: [localhost]

TASK [pgo-operator : Check PGO Service Account] ********************************
changed: [localhost]

TASK [pgo-operator : Create PGO Service Account] *******************************
skipping: [localhost]

TASK [pgo-operator : Template Cluster RBAC (namespace_mode 'dynamic')] *********
changed: [localhost]

TASK [pgo-operator : Check Cluster RBAC (namespace_mode 'dynamic')] ************
changed: [localhost]

TASK [pgo-operator : Create Cluster RBAC (namespace_mode 'dynamic')] ***********
skipping: [localhost]

TASK [pgo-operator : Template Cluster RBAC (namespace_mode 'readonly')] ********
skipping: [localhost]

TASK [pgo-operator : Check Cluster RBAC (namespace_mode 'readonly')] ***********
skipping: [localhost]

TASK [pgo-operator : Create Cluster RBAC (namespace_mode 'readonly')] **********
skipping: [localhost]

TASK [pgo-operator : Cluster Roles Disabled (namespace_mode 'disabled')] *******
skipping: [localhost]

TASK [pgo-operator : Create PGO Image Pull Secret] *****************************
skipping: [localhost]

TASK [pgo-operator : Check cluster-admin Cluster Role Binding for PGO Service Account] ***
skipping: [localhost]

TASK [pgo-operator : Create cluster-admin Cluster Role Binding for PGO Service Account] ***
skipping: [localhost]

TASK [pgo-operator : Template PGO RBAC] ****************************************
changed: [localhost]

TASK [pgo-operator : Check PGO RBAC] *******************************************
changed: [localhost]

TASK [pgo-operator : Create PGO RBAC] ******************************************
skipping: [localhost]

TASK [pgo-operator : Template Local PGO User] **********************************
changed: [localhost]

TASK [pgo-operator : Check PGO BackRest Repo Secret] ***************************
changed: [localhost]

TASK [pgo-operator : Create PGO BackRest Repo Secret] **************************
skipping: [localhost]

TASK [pgo-operator : Check PGO API Secret] *************************************
changed: [localhost]

TASK [pgo-operator : Create PGO API Secret] ************************************
skipping: [localhost]

TASK [pgo-operator : Template PGO Configuration] *******************************
changed: [localhost]

TASK [pgo-operator : Check PGO ConfigMap] **************************************
changed: [localhost]

TASK [pgo-operator : Create PGO ConfigMap] *************************************
skipping: [localhost]

TASK [pgo-operator : Template PGO Service Configuration] ***********************
changed: [localhost]

TASK [pgo-operator : Check PGO Service Configuration] **************************
changed: [localhost]

TASK [pgo-operator : Create PGO Service] ***************************************
skipping: [localhost]

TASK [pgo-operator : Template PGO Deployment] **********************************
changed: [localhost]

TASK [pgo-operator : Check PGO Deployment] *************************************
changed: [localhost]

TASK [pgo-operator : Deploy PGO] ***********************************************
changed: [localhost]

TASK [pgo-operator : Wait for PGO to finish deploying] *************************
fatal: [localhost]: FAILED! => {"ansible_job_id": "936501454956.1054", "changed": true, "cmd": ["kubectl", "rollout", "status", "deployment/postgres-operator", "-n", "pgo"], "delta": "0:09:57.166834", "end": "2020-09-21 05:23:56.014882", "finished": 1, "msg": "non-zero return code", "rc": 1, "start": "2020-09-21 05:13:58.848048", "stderr": "error: deployment \"postgres-operator\" exceeded its progress deadline", "stderr_lines": ["error: deployment \"postgres-operator\" exceeded its progress deadline"], "stdout": "Waiting for deployment \"postgres-operator\" rollout to finish: 0 of 1 updated replicas are available...", "stdout_lines": ["Waiting for deployment \"postgres-operator\" rollout to finish: 0 of 1 updated replicas are available..."]}

PLAY RECAP *********************************************************************
localhost                  : ok=52   changed=33   unreachable=0    failed=1    skipped=37   rescued=0    ignored=0   
```
I have other storage classes defined but they shouldn't have a bearing on the quickstart unless I start editing values.yaml. This fails in my environment before I touch anything.
```
$ kubectl get storageclass
NAME                      PROVISIONER                    AGE
local-storage (default)   kubernetes.io/no-provisioner   438d
nfs-client                kubernetes.io/no-provisioner   438d
```

**Comment by jkatz** [comment]
> I'm a newbie so this may be a repeat request--but it shouldn't take me two hours of struggling to conclude that, basically, the quickstart isn't quick and requires a bug report.

The quickstart also comes with a disclaimer:

<img width="1168" alt="Screen Shot 2020-09-21 at 8 59 45 AM" src="https://user-images.githubusercontent.com/1694/93769457-cd165a00-fbe8-11ea-8dba-b01a08816b5e.png">

**If your environment is set up to use hostpath storage (found in things like minikube or OpenShift Code Ready Containers, the following command could work for you:**

**If not, please read onward: you can still get up and running fairly quickly with just a little bit of configuration.**

And as such, you may need to take additional steps. Based on the description of your environment, it sounds like you need to make modifications to your defined storage configurations (which is covered in the quick start):

[https://access.crunchydata.com/documentation/postgres-operator/4.4.1/quickstart/#configure-the-postgresql-operator-installer](https://access.crunchydata.com/documentation/postgres-operator/4.4.1/quickstart/#configure-the-postgresql-operator-installer)

which also includes a link that expands upon how they are defined:

[https://access.crunchydata.com/documentation/postgres-operator/4.4.1/installation/configuration/#storage-configuration-options](https://access.crunchydata.com/documentation/postgres-operator/4.4.1/installation/configuration/#storage-configuration-options)

Reading your error, it sounds like you set `primary_storage` to be the value `dynamic`, which does not match any of the storage configuration names you set. It appears you have dynamic provisioning for your hostpath and NFS storage, so perhaps try something like:

```yaml
backrest_storage: "local"
backup_storage: "local"
primary_storage: "local"
replica_storage: "local"
# ...
storage1_name: "local"
storage1_access_mode: "ReadWriteMany"
storage1_size: "10Gi"
storage1_type: "dynamic"
storage1_class: "local-storage"
```

I do not know if the above works as I do not have those provisioners installed, but I would recommend trying that out. You _might_ need to add a supplemental group to it (e.g. storage1_supplemental_groups: "65534") but without knowing more specifics about your environment and provisioners I can't comment to that.

Lastly, it's really difficult to write a one-size-fits-all quickstart for things that involve persistence in Kubernetes: there are many different permutations of Kubernetes versions, distributions storage classes etc., the goal with any quickstart I've seen is to provide enough information to get started in many base cases, followed by links to documentation if one gets stuck.

This is similar to how PostgreSQL sets its default configurations as well: there are many different machines that PostgreSQL could be deployed to. PostgreSQL's defaults are actually very small, e.g. you would not want to run with `shared_buffers=512MB` on a machine with 32GB of RAM -- you would not get many of the benefits of that!

However, it does allow for you to get started and ultimately the documentation guides on how to set up PostgreSQL to run in your environment.

**Comment by instantlinux** [comment]
"it sounds like you set primary_storage to be the value dynamic". Maybe I did during an earlier experiment, but when I posted this issue, I'm just doing an `apply` of the published resource definitions in which I am not "setting" anything. Defaults for `primary_storage` and related items in values.yaml are hostpathstorage. Where is this value "dynamic" coming from? Could there still be a stale value that I have not cleared out between my experiements? Is it enough to delete the employments and the other items found by this command?
```
$ kubectl get all -n pgo --context=sudo
NAME                                     READY   STATUS             RESTARTS   AGE
pod/pgo-deploy-42s4k                     1/1     Running            0          12m
pod/postgres-operator-66cc8bd589-w666m   1/4     CrashLoopBackOff   15         4m55s


NAME                        TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)                      AGE
service/postgres-operator   ClusterIP   10.108.207.69   <none>        8443/TCP,4171/TCP,4150/TCP   18h

NAME                                READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/postgres-operator   0/1     1            0           4m55s

NAME                                           DESIRED   CURRENT   READY   AGE
replicaset.apps/postgres-operator-66cc8bd589   1         1         0       4m55s

NAME                   COMPLETIONS   DURATION   AGE
job.batch/pgo-deploy   0/1           12m        12m
```

**Comment by jkatz** [comment]
Can you please post the current value of the `pgo.yaml` in the `pgo-config` YAML? You could find it via

`kubectl edit cm -n pgo pgo-config`

**Comment by instantlinux** [comment]
This is what I see:
```
  pgo.yaml: "Cluster:\n  CCPImagePrefix:  registry.developers.crunchydata.com/crunchydata\n
    \ CCPImageTag:  centos7-12.4-4.4.1\n  DisableAutofail:  false\n  BackrestPort:
    \ 2022\n  BackrestS3Bucket: \n  BackrestS3Endpoint: \n  BackrestS3Region: \n  BackrestS3URIStyle:
    \n  BackrestS3VerifyTLS: \"true\"\n  Metrics:  false\n  Badger:  false\n  Port:
    \ 5432\n  PGBadgerPort: 10000\n  ExporterPort: 9187\n  User:  testuser\n  Database:
    \ \n  PasswordAgeDays:  0\n  PasswordLength:  24\n  Replicas:  0\n  ArchiveMode:
    \ true\n  ServiceType:  ClusterIP\n  EnableCrunchyadm:  false\n  DisableReplicaStartFailReinit:
    \ false\n  PodAntiAffinity:  preferred\n  PodAntiAffinityPgBackRest: \n  PodAntiAffinityPgBouncer:
    \n  SyncReplication:  false\n  DefaultInstanceMemory: 128Mi\n  DefaultBackrestMemory:
    48Mi\n  DefaultPgBouncerMemory: 24Mi\n  DisableFSGroup: false\nPrimaryStorage:
    dynamic\nWALStorage: \nBackupStorage: hostpathstorage\nReplicaStorage: dynamic\nBackrestStorage:
    hostpathstorage\nStorage:\n  hostpathstorage:\n    AccessMode:  ReadWriteMany\n
    \   Size:  1G\n    StorageType:  create\n  replicastorage:\n    AccessMode:  ReadWriteMany\n
    \   Size:  1G\n    StorageType:  create\n  nfsstorage:\n    AccessMode:  ReadWriteMany\n
    \   Size:  1G\n    StorageType:  create\n    SupplementalGroups: 65534\n  nfsstoragered:\n
    \   AccessMode:  ReadWriteMany\n    Size:  1G\n    StorageType:  create\n    MatchLabels:
    crunchyzone=red\n    SupplementalGroups: 65534\n  storageos:\n    AccessMode:
    \ ReadWriteOnce\n    Size:  5Gi\n    StorageType:  dynamic\n    StorageClass:
    fast\n  primarysite:\n    AccessMode:  ReadWriteOnce\n    Size:  4G\n    StorageType:
    \ dynamic\n    StorageClass: primarysite\n  alternatesite:\n    AccessMode:  ReadWriteOnce\n
    \   Size:  4G\n    StorageType:  dynamic\n    StorageClass: alternatesite\n  gce:\n
    \   AccessMode:  ReadWriteOnce\n    Size:  300M\n    StorageType:  dynamic\n    StorageClass:
    standard\n  rook:\n    AccessMode:  ReadWriteOnce\n    Size:  1Gi\n    StorageType:
    \ dynamic\n    StorageClass: rook-ceph-block\nPgo:\n  Audit:  false\n  DisableReconcileRBAC:
    False\n  PGOImagePrefix:  registry.developers.crunchydata.com/crunchydata\n  PGOImageTag:
    \ centos7-4.4.1\n"
```
How do I clobber that? Can the doc make this more clear? I think the reason there are several tickets similar to this one is that the docs are very thin on a) how to reset the config and start over; b) what pgo.yaml is and how it relates to values.yaml; and c) how to yank out the default storage settings and replace them with your own.

**Comment by jkatz** [comment]
If you are just reapplying the configuration, the original configuration will remain. You would have to uninstall. This is covered here:

[https://access.crunchydata.com/documentation/postgres-operator/4.4.1/installation/postgres-operator/#uninstalling](https://access.crunchydata.com/documentation/postgres-operator/4.4.1/installation/postgres-operator/#uninstalling)

For "overwriting with your own storage" settings, this is covered both in the documentation I linked you above, as well as the more thorough installation instructions:

[https://access.crunchydata.com/documentation/postgres-operator/4.4.1/installation/postgres-operator/](https://access.crunchydata.com/documentation/postgres-operator/4.4.1/installation/postgres-operator/)
[https://access.crunchydata.com/documentation/postgres-operator/4.4.1/installation/configuration/](https://access.crunchydata.com/documentation/postgres-operator/4.4.1/installation/configuration/)

pgo.yaml has its own section where it is covered in the documentation: [https://access.crunchydata.com/documentation/postgres-operator/4.4.1/configuration/pgo-yaml-configuration/](https://access.crunchydata.com/documentation/postgres-operator/4.4.1/configuration/pgo-yaml-configuration/). These are the template variables and as well as the overall configuration for the Operator. `values.yaml` is used only as part of the deploy process, where as `pgo.yaml` persists beyond that, as the Operator reads its default settings and configurations from there.

With the default storage settings, the documentation is up front about this in the quick start as reference [here](https://access.crunchydata.com/documentation/postgres-operator/4.4.1/quickstart/#configure-the-postgresql-operator-installer). A bunch of the documentation was written at different points and time and could always continued to be improved. 

The PostgreSQL Operator project has been around for awhile and so too has the documentation. While it has evolved through the years, the documentation tends to be worked on when things are added. Occasionally holistic passes are made to help with the overall flow of the documentation, but I am confident this can continued to be improved.

IF you can point to specific areas of the documentation that can be improved or contribute patches, that of course would be welcome.

Thanks!

**Comment by instantlinux** [comment]
I have gotten up and running, thanks for the help! My perspective is that of a long-timer in k8s (you can see I'm still running 1.15!) who has been running MySQL/MariaDB at expert level for the past couple of decades (enough that I even authored a cluster-builder, see the one I still maintain here at [/instantlinux/dockertools/tree/master/images/mariadb-galera]) but has only run single-instance postgresql here and there for apps that require it. I vaguely knew of Zalando in the past but never heard of CrunchyData until I stumbled into it this week. It looks like quickstart is a recent addition, which makes `pgo` a lot easier to set up than it used to be--congrats on that. But remember that people drawn to anything labeled "quickstart" are almost by definition unfamiliar with the past issues faced by CrunchyData developers and users.

So, here's my $0.02 to improve it and help the next quickstart user.

* There is a command `kubectl get all -n pgo` which doesn't quite show everything, in particular ConfigMaps.
* Tools like chef/puppet developed the concept of idempotency over the past couple of decades, so the `kubectl apply` command is based on the essentials of idempotency; simply put, if you edit the input file and invoke that command one or more times, the runtime system should pick up all changes you make and not make changes you didn't make
* Specifically, ask the developers how `values.yaml` can be read at startup time by the operator or deploy image to apply changes seen in it; that way a simple pod restart should be all that's needed
* Ansible got your deployer developers up and running quickly, but it's notoriously slow for users and in test pipelines (ansible configurations grow more complex each release): mine takes about 7 minutes to run--consider refactoring many of the steps it makes into python-native code invoked as custom ansible modules
* In the [quickstart page](https://access.crunchydata.com/documentation/postgres-operator/latest/quickstart/), please add a troubleshooting section on the same page; in that section, spell out how to uninstall, how to reset the configs (just `kubectl delete cm` for any listed by `kubectl get cm`), and where to find logs or other helpful info that I haven't discovered yet
* Until there is a way for `values.yaml` updates to be applied to a running system, it looks to me that quickstart isn't complete; add a disclaimer that the quickstart is helpful for setting up a demo-quality installation but that for anything beyond that a user will have to learn other ways to make any changes one might make to `values.yaml` - which are documented in other places and involve changes to files other than the quickstart's yaml

OK, my input might only be worth $0.01 -- hopefully it's helpful -- thanks again!

**Comment by jkatz** [comment]
@instantlinux I will review all the points later. I was also curious if you would accept #1912 as part of the solution, i.e. treating the empty storage class as using the Kubernetes default? I think that may truly make the "super quick start" the actual quick start, and save a lot of pain in the future :wink:

**Comment by jkatz** [comment]
@instantlinux Thanks for the suggestions. We were able to employ some of them before the v4.5, notably adding the default storage classes as mentioned.

We also expanded the [quickstart](https://access.crunchydata.com/documentation/postgres-operator/latest/quickstart/) and added a [tutorial](https://access.crunchydata.com/documentation/postgres-operator/latest/tutorial/), each also containing sections on troubleshooting. The tutorial certainly needs some more work, but the aim is to provide a better walkthrough of the daily tasks.

The addition of being able to use the default storage class (and making that the default Operator configuration) should simplify a lot of the issues mention.

Hope this helpS!
